dipy/core/sphere.py

```diff
@@ -498,6 +500,155 @@ def disperse_charges(hemi, iters, const=.2):
+    weights : array-like (N, N)
+        Weight values to the electrostatic energy.
+
+    Returns
+    -------
+    energy : float
+        Sum of all interactions between any two vectors.
+
+    """
+
+    nb_points = vects.shape[0] / 3
+    # weights = kwargs.get('weights', np.ones((nb_points, nb_points)) / nb_points**2)
+    weights = kwargs.get('weights', np.ones((nb_points, nb_points)))
+    charges = vects.reshape((nb_points, 3))
+    all_charges = np.concatenate((charges, -charges))
+    all_charges = all_charges[:, None]
+    r = charges - all_charges
+    r_mag = np.sqrt((r * r).sum(-1))[:, :, None]
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
+        potential = 1 / r_mag ** alpha
+
+    d = np.arange(len(charges))
+    potential[d, d] = 0
+    potential = potential[:nb_points] + potential[nb_points:]
+    potential = weights * potential.sum(-1)
+    potential = potential.sum()
+    return potential
+
+
+def _get_grad_forces_alt(vects, alpha=2.0, **kwargs):
+    """1st-order derivative of electrostatic-like repulsion energy.
+
+    Parameters
+    ----------
+    vects : array-like (N * 3,)
+        Points on the sphere.
+    alpha : float
+        Controls the power of the repulsion. Default is 1.0.
+
+    Returns
+    -------
+    grad : array-like (N * 3,)
+        Gradient of the objective function.
+
+    """
+
+    nb_points = vects.shape[0] / 3
+    # weights = kwargs.get('weights', np.ones((nb_points, nb_points)) / nb_points**2)
+    weights = kwargs.get('weights', np.ones((nb_points, nb_points)))
+    charges = vects.reshape((nb_points, 3))
+    all_charges = np.concatenate((charges, -charges))
+    all_charges = all_charges[:, None]
+    r = charges - all_charges
+    r_mag = np.sqrt((r * r).sum(-1))[:, :, None]
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
+        forces = -2 * alpha * r / r_mag ** (alpha + 2.)
+
+    d = np.arange(len(charges))
+    forces[d, d] = 0
+    forces = forces[:nb_points] + forces[nb_points:]
+    forces = forces * weights.reshape((nb_points, nb_points, 1))
+    forces = forces.sum(0)
+    return forces.reshape((nb_points * 3))
+
+
+def disperse_charges_alt(init_pointset, iters, tol=1.0e-3):
+    """Reimplementation of disperse_charges making use of
+    scipy.optimize.fmin_slsqp.
+
+    Parameters
+    ----------
+    init_pointset : HemiSphere
+        Points on a unit sphere.
+    iters : int
+        Number of iterations to run.
+    tol : float
+        Tolerance for the optimization.
+
+    Returns
+    -------
+    array-like (N * 3,)
+        Distributed points on a unit sphere.
+
+    """
+
+    K = init_pointset.shape[0]
+    vects = optimize.fmin_slsqp(_get_forces_alt, init_pointset.reshape(K * 3),
+                                f_eqcons=_equality_constraints,
+                                fprime=_get_grad_forces_alt, iter=iters, acc=tol,
+                                args=[], iprint=0)
+    return vects
```

Comment:
Here you may want to return vects.reshape((K, 3))

Reply:
Not sure. I guess `optimize.fmin_slsqp` returns an array that has the same shape as
its `x0` parameter (`init_pointset.reshape(K * 3)` in this case)? Am I wrong?

Reply:
> Not sure. I guess `optimize.fmin_slsqp` returns an array that has the same shape as its `x0` parameter (`init_pointset.reshape(K * 3)` in this case)? Am I wrong?

Yes, I think we reshape `x0` to a 1D array because `scipy.optimize` requires this. Don't you think it would be convenient to return here a reshaped array, since we are dealing with 3d points in space?


> I had not considered adding some profiling. I have not used profiling tools so far, but I would be willing to. I think it is a valuable addition. So if you have pointers, I'd be grateful.

I am not an expert in profiling, either... let's wait for suggestions by @arokem


> Also, I guess we should have these methods tested/exercised, right? Do you have any proposal for this?

A simple test I can think of is:
1. generate a random set of points, 
2. compute the associated electrostatic potential
3. disperse charges
4. verify that the potential of the optimal configuration is smaller than that of the original configuration

Reply:
Thanks for the follow-up @ecaruyer !

> Yes, I think we reshape x0 to a 1D array because scipy.optimize requires this. Don't you think it would be convenient to return here a reshaped array, since we are dealing with 3d points in space?

Agreed. Done in 5e3bd7d.

>  Also, I guess we should have these methods tested/exercised, right? Do you have any proposal for this?
A simple test I can think of is:
...

Please check if 989b512 is according to expected.